### PR TITLE
Add Publisher#cast, Single#cast, and Publisher#ofType

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
Filtering and casting based upon types maybe a common
use case but it currently requires knowledge of Class
methods and potentially multiple operators.

Modifications:
- Add Publisher#cast, Single#cast, and Publisher#ofType
  operators and associated tests.

Result:
Operators exist which make casting and filtering between
types easier.

### CODE REVIEW

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java [hunk 1]
```diff
@@ -118,6 +118,26 @@ protected Publisher() {
...
+     *         results.add(clazz.cast(t));
+     *     }
+     *     return results;
+     * }</pre>
+     * @param clazz The type to cast to.
+     * @param <R> The resulting type of the cast operation.
+     * @return The cast of this {@link Publisher} to type {@link R}.
+     * @see <a href="https://reactivex.io/documentation/operators/map.html">ReactiveX cast operator.</a>
+     */
+    public final <R> Publisher<R> cast(Class<R> clazz) {
```
Comment (Scottmitch) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java:137]: these operators are not impossible for users todo them selves but are commonly provided by other RS implementations and lowers the barrier to entry.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java [hunk 2]
```diff
@@ -134,13 +154,34 @@ protected Publisher() {
...
      * @return A {@link Publisher} that only emits the items that pass the {@code predicate}.
-     *
      * @see <a href="http://reactivex.io/documentation/operators/filter.html">ReactiveX filter operator.</a>
      */
     public final Publisher<T> filter(Predicate<? super T> predicate) {
         return filter(() -> predicate);
     }
 
+    /**
+     * Filters items so that only items of type {@link R} are emitted by the return value.
```
Comment (bondolo) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java:164]: …only non-null items of…

**Review by bondolo — COMMENTED** [review]
I like these operators as using Java streams I frequently use the `filter(clazz::isInstance).map(clazz::cast)` pattern.

I wonder how we should document the effects of casting a bad item with `cast()` so that people can add their appropriate `onErrorResume` for the `ClassCastException`s which may result.

Also worth documenting that `ofType` will filter nulls.

File: servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherOfTypeTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        assertThat(subscriber.takeOnNext(), is(1));
+        source.onComplete();
+        subscriber.awaitOnComplete();
+    }
+
+    @Test
+    void noSignals() {
+        toSource(source.ofType(Integer.class)).subscribe(subscriber);
+        subscriber.awaitSubscription().request(2);
+        source.onNext("NotInteger1", "NotInteger2");
```
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherOfTypeTest.java:55]: Consider adding a test case with `null` item

**Review by idelpivnitskiy — APPROVED** [review]
LGTM after @bondolo comments addressed

**Comment by Scottmitch** [comment]
build failure attributed to https://github.com/apple/servicetalk/issues/2036

**Review by bondolo — APPROVED** [review]
